Fix NetworkRTCMonitor gatherNetworkMap

youennf opened the pull request:
#### 926c05cb3c2a6cb7804f902b8680050d11a59e37
<pre>
Fix NetworkRTCMonitor gatherNetworkMap
<a href="https://bugs.webkit.org/show_bug.cgi?id=277329">https://bugs.webkit.org/show_bug.cgi?id=277329</a>
<a href="https://rdar.apple.com/132591641">rdar://132591641</a>

Reviewed by Chris Dumez.

The utf8 string may be deallocated before the span, we are now using the span within RTCNetwork constructor.
I also use more spans to reduce string allocation.

* Source/WebKit/NetworkProcess/webrtc/NetworkRTCMonitor.cpp:
(WebKit::gatherNetworkMap):

Canonical link: <a href="https://commits.webkit.org/281618@main">https://commits.webkit.org/281618@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c3c021762839c54e14772cbd0387b88e52c741c0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/60375 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/39726 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/12933 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/64294 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/10906 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/47398 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/11139 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/48874 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7593 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/62406 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/37021 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/52306 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29721 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/33719 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/9823 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/55601 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/9820 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/66026 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/4308 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/9656 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/56234 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/4327 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/52279 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/56404 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/3589 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9085 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/35536 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/36618 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/37707 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/36362 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/NetworkProcess/webrtc/NetworkRTCMonitor.cpp:
```diff
@@ -215,7 +215,8 @@ static HashMap<String, RTCNetwork> gatherNetworkMap()
             std::span<const char> name { iterator->ifa_name, strlen(iterator->ifa_name) };
-            std::span<const char> description { networkKey.utf8().data(), networkKey.utf8().length() };
+            auto descriptionString = networkKey.utf8();
+            std::span<const char> description { descriptionString.data(), descriptionString.length() };
```
We can drop this...

cdumez commented on Source/WebKit/NetworkProcess/webrtc/NetworkRTCMonitor.cpp:
```diff
@@ -215,7 +215,8 @@ static HashMap<String, RTCNetwork> gatherNetworkMap()
-            std::span<const char> description { networkKey.utf8().data(), networkKey.utf8().length() };
+            auto descriptionString = networkKey.utf8();
+            std::span<const char> description { descriptionString.data(), descriptionString.length() };
             return RTCNetwork { { name }, { description }, address->second, prefixLength, interfaceAdapterType(name), 0, 0, true, false, scopeID, { } };
```
... and pass `description.span()` here.

youennf replied:
Right, I'll use span here and for name as well.

cdumez commented on Source/WebKit/NetworkProcess/webrtc/NetworkRTCMonitor.cpp:
```diff
@@ -215,7 +215,8 @@ static HashMap<String, RTCNetwork> gatherNetworkMap()
         networkMap.ensure(networkKey, [&] {
             std::span<const char> name { iterator->ifa_name, strlen(iterator->ifa_name) };
-            std::span<const char> description { networkKey.utf8().data(), networkKey.utf8().length() };
+            auto descriptionString = networkKey.utf8();
```
nit: can keep `description` as variable name with my suggestion below.

cdumez commented on Source/WebKit/NetworkProcess/webrtc/NetworkRTCMonitor.cpp:
```diff
@@ -215,7 +215,8 @@ static HashMap<String, RTCNetwork> gatherNetworkMap()
 
         networkMap.ensure(networkKey, [&] {
             std::span<const char> name { iterator->ifa_name, strlen(iterator->ifa_name) };
```
`auto name = span(iterator->ifa_name);`

cdumez reviewed (approved):
LGTM with nits.

youennf committed:
Fix NetworkRTCMonitor gatherNetworkMap
https://bugs.webkit.org/show_bug.cgi?id=277329
rdar://132591641

Reviewed by Chris Dumez.

The utf8 string may be deallocated before the span, we are now using the span within RTCNetwork constructor.
I also use more spans to reduce string allocation.

* Source/WebKit/NetworkProcess/webrtc/NetworkRTCMonitor.cpp:
(WebKit::gatherNetworkMap):

Canonical link: https://commits.webkit.org/281618@main
